src/transformers/modeling_tf_utils.py

```diff
@@ -753,6 +754,39 @@ def get_input_embeddings(self) -> tf.keras.layers.Layer:
         else:
             raise NotImplementedError
 
+    def _save_checkpoint(self, checkpoint_dir, epoch):
+        if not os.path.isdir(checkpoint_dir):
+            os.mkdir(checkpoint_dir)
+        # We avoid tf.train.checkpoint or saving weights in TF format, even though that includes optimizer
+        # state for us, because it requires special handling for objects like custom losses, which we use
+        # internally and which users are likely to use too
+        weights_path = os.path.join(checkpoint_dir, "weights.h5")
+        self.save_weights(weights_path)
+        extra_data = {"epoch": epoch, "optimizer_state": self.optimizer.get_weights()}
+        extra_data_path = os.path.join(checkpoint_dir, "extra_data.pickle")
+        with open(extra_data_path, "wb") as f:
+            pickle.dump(extra_data, f)
+
+    def load_repo_checkpoint(self, repo_path_or_name, organization=None):
+        if getattr(self, "optimizer", None) is None:
+            raise RuntimeError(
+                "Checkpoint loading failed as no optimizer is attached to the model. "
+                "This is most likely caused by the model not being compiled."
+            )
+        repo = self._create_or_get_repo(repo_path_or_name, organization)
```

Comment:
I think we said to use `Repository` here to avoid creating a new repo?

Reply:
I tried this, but it resulted in some code duplication in order to make everything work - I kept it as-is but added a comment to explain instead. If you'd still prefer to avoid the call, let me know!

Reply:
Let's see what @LysandreJik thinks then.

Reply:
Are you sure it won't actually create a remote repository if it doesn't exist? If you're certain of it it's fine for me to go with `self._create_or_get_repo` but I'm quite unsure that it won't create a remote repo.

Also which method are you using this method here?

https://github.com/huggingface/transformers/blob/956a483173e77ebf655ca9636a5f7b6ef010b307/src/transformers/file_utils.py#L2338-L2367

Because if so, don't call it with positional arguments like you do here as you're passing the `organization` parameter to `repo_url`.

Reply:
Good spot on the `organization` positional arg, fixed!

Also, the `_create_or_get_repo` method just does this, after sanity-checking the arguments:

```
repo = Repository(repo_path_or_name, clone_from=repo_url, use_auth_token=use_auth_token)
repo.git_pull()
return repo
```
So I think as long as we don't `push` it afterwards, we won't create a new remote repo if one doesn't exist already.

Reply:
I think this discussion is real proof we need to just put the code used here and not try to use some magic of the Mixin. The library always prefers an explicit approach at the cost of duplicate code, and I think this is an instance where we should just apply the code needed.

The Mixin might also change in the feature, following the RFC on huggingface_hub, so those private methods might actually introduce some behavior we don't want in the future.

Reply:
And the `Repository(repo_path_or_name, clone_from=repo_url, use_auth_token=use_auth_token)` line actually creates the repo if you try to clone it but it doesn't exist:

https://github.com/huggingface/huggingface_hub/blob/10c69146969ad7f9e1add075c1ef4ec15e42e85f/src/huggingface_hub/repository.py#L542-L549

Reply:
The `push` event is unrelated to the creation of the repository. No push to a remote repository can be done if the repository does not exist. The repository creation happens at the repository creation, not at the repository update (push).

Reply:
Ah, I'm sorry, I didn't realize that's how it worked! And yes, you're right, this is probably a sign that we need to be explicit. I'll try to figure out some code that just checks without accidentally creating a new repo.